src/psyclone/domain/lfric/lfric_arg_descriptor.py

```diff
@@ -0,0 +1,676 @@
+
+    @property
+    def function_spaces(self):
+        '''
+        Returns the function space names that this instance operates on as
+        a list depending on the argument type: one function space for a
+        field, both function spaces for an operator and an empty list for
+        a scalar.
+
+        :returns: function spaces than an argument instance operates on.
+        :rtype: list of str
+
+        :raises InternalError: if an invalid argument type is passed in.
+
+        '''
+        if self._type in LFRicArgDescriptor.VALID_FIELD_NAMES:
+            return [self.function_space]
+        if self._type in LFRicArgDescriptor.VALID_OPERATOR_NAMES:
+            # return to before from to maintain expected ordering
+            return [self.function_space_to, self.function_space_from]
+        if self._type in LFRicArgDescriptor.VALID_SCALAR_NAMES:
+            return []
+        raise InternalError("LFRicArgDescriptor.function_spaces(), should "
+                            "not get to here.")
+
+    @property
+    def vector_size(self):
+        '''
+        Returns the vector size of the argument. This will be 1 if *n
+        has not been specified for all arguments except scalars (their
+        vector size is set to 0).
+
+        :returns: vector size of the argument.
+        :rtype: str
+
+        '''
+        return self._vector_size
+
+    def __str__(self):
+        '''
+        Creates a string representation of the argument descriptor. This
+        is type and access for scalars with the addition of function
+        space(s) for fields and operators.
+
+        :returns: string representation of the argument descriptor.
+        :rtype: str
+
+        :raises InternalError: if an invalid argument type is passed in.
+
+        '''
+        res = "LFRicArgDescriptor object" + os.linesep
+        res += "  argument_type[0]='{0}'".format(self._type)
+        if self._vector_size > 1:
+            res += "*"+str(self._vector_size)
+        res += os.linesep
+        res += "  access_descriptor[1]='{0}'"\
+               .format(self._access_type.api_specific_name())\
+               + os.linesep
+        if self._type in LFRicArgDescriptor.VALID_FIELD_NAMES:
+            res += "  function_space[2]='{0}'".format(self._function_space1) \
+                   + os.linesep
+        elif self._type in LFRicArgDescriptor.VALID_OPERATOR_NAMES:
+            res += "  function_space_to[2]='{0}'".\
+                   format(self._function_space1) + os.linesep
+            res += "  function_space_from[3]='{0}'".\
+                   format(self._function_space2) + os.linesep
+        elif self._type in LFRicArgDescriptor.VALID_SCALAR_NAMES:
+            pass  # we have nothing to add if we're a scalar
+        else:  # we should never get to here
+            raise InternalError("LFRicArgDescriptor.__str__(), should not "
+                                "get to here.")
+        return res
```

Comment:
Are you sure about that? I followed the logic set out in the `FunctionSpace` refactoring and added `LFRicArgDescriptor` in the `__init__` module rather than here. `function_space.py` also does not have the list in the module.